exonum/src/runtime/rust/service.rs

```diff
@@ -170,7 +169,20 @@ impl<'a, 'b> TransactionContext<'a, 'b> {
     #[doc(hidden)]
     /// Create a context to call interfaces of the specified service instance.
     pub fn call_context(&self, called: InstanceId) -> CallContext<'a> {
-        CallContext::new(self.inner, self.instance.id, called)
+        CallContext::from_execution_context(self.inner, self.instance.id, called)
+    }
+
+    /// Check the caller of this method with the specified closure.
+    ///
+    /// If the closure returns `Ok(value)`, then the method returns `Ok((value, fork))` thus you
+    /// get a write access to the blockchain state. Otherwise this method returns
+    /// an occurred error.
+    pub fn check_caller<F, T, E>(&self, check: F) -> Result<(T, &Fork), E>
```

Comment:
Does it really have to be that complicated?..

Reply:
Maybe no, this is just only an experiment in the public API.
But I believe that we must protect service writers from the easy footshots. If you know how I can simplify this code just say.

Reply:
`F` can be `Fn(&Caller) -> bool`, so we get rid of the `T` and `E` and also return just `&Fork` instead of `Result` with tuple.
Also `check` can be renamed to `predicate`, because context is already telling you that `check_caller` performs some kind of check, and `predicate` will be indicating that it's a function.

Reply:
I think that information like caller instance ID or transaction author is very useful as the result of check.

Reply:
Well, yes, but by introducing such a functionality, the function scope goes far from just `check_caller`. Nothing prevents user from wtiting code like

```rust
if let Some(fork) = context.check_caller(some_predicate) {
    let caller_info = extract_data_from_caller(context.caller());
}
```

IMHO It will be even more straightforward than
```
let (fork, caller_info) = context.check_caller(some_function)?; // Why `check_caller` returns some info?
```

Reply:
In this case extraction if information performs twice.